File: src/main/java/javaclasses/Recursion/Permutations.java
/*
 * * Permutations.java
 *  * Created by Rafsan Ahmad on 10/25/21, 9:49 PM
 *  * Copyright (c) 2021 . All rights reserved.
 *
 */

package javaclasses.Recursion;

import java.util.ArrayList;
import java.util.List;

//https://leetcode.com/problems/permutations
//Leetcode 46
/*
* Given an array nums of distinct integers, return all the possible permutations.
* You can return the answer in any order.
*
Example 1:
Input: nums = [1,2,3]
Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]

Example 2:
Input: nums = [0,1]
Output: [[0,1],[1,0]]

Example 3:
Input: nums = [1]
Output: [[1]]*/
public class Permutations {
    //res/permutation_tree.jpg
    //res/permutation_tree_2.jpg

    private void recurPermute(int[] nums, List<Integer> perms, List<List<Integer>> ans, boolean[] freq) {
        if (perms.size() == nums.length) {
            ans.add(new ArrayList<>(perms));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (!freq[i]) {
                freq[i] = true;
                perms.add(nums[i]);
                recurPermute(nums, perms, ans, freq);
                perms.remove(perms.size() - 1);
                freq[i] = false;
            }

        }
    }

    //Using Frequency array
    /*
     * Time Complexity: O(N! X N)
     * Space Complexity: O(N)
     * */
    public List<List<Integer>> permuteUsingFrequency(int[] nums) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> perms = new ArrayList<>();
        boolean[] freq = new boolean[nums.length];
        recurPermute(nums, perms, ans, freq);
        return ans;
    }

    // Using memory efficient way
    /*
     * Time Complexity: O(N! X N)
     * Space Complexity: O(1)
     * */
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        recurPermute(0, nums, result);
        return result;
    }

    private void recurPermute(int start, int[] nums, List<List<Integer>> result) {
        if (start == nums.length - 1) {
            List<Integer> list = new ArrayList<>();
            for (int num : nums) {
                list.add(num);
            }
            result.add(list);
            return;
        }

        for (int i = start; i < nums.length; i++) {
            swap(nums, i, start);
            recurPermute(start + 1, nums, result);
            swap(nums, i, start);
        }
    }

    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    public static void main(String[] args) {
        Permutations perm = new Permutations();
        int[] arr = {1, 2, 3};
        List<List<Integer>> result = perm.permute(arr);
        System.out.println(result);

        List<List<Integer>> result2 = perm.permuteUsingFrequency(arr);
        System.out.println(result2);
    }

}

File: src/main/java/kotlinclasses/Backtracking/Permutations.kt
/*
 * *
 *  * Permutations.kt
 *  * Created by Rafsan Ahmad on 6/27/25, 9:51PM
 *  * Copyright (c) 2025 . All rights reserved.
 *  *
 *
 */

package kotlinclasses.Backtracking

class Permutations {
    //https://leetcode.com/problems/permutations/description/
    /*Given an array nums of distinct integers, return all the possible permutations.
    You can return the answer in any order.

Example 1:
Input: nums = [1,2,3]
Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]

Example 2:
Input: nums = [0,1]
Output: [[0,1],[1,0]]

Example 3:
Input: nums = [1]
Output: [[1]]


Constraints:
1 <= nums.length <= 6
-10 <= nums[i] <= 10
All the integers of nums are unique.*/

    /*Time Complexity:
🔸 O(n × n!)
Space Complexity:
🔸 O(n × n!) for the output
🔸 O(n) auxiliary for recursion + temp structures

Reason:
There are n! permutations of a list with n unique elements.
For example, for [1,2,3]: total permutations = 3! = 6
For each permutation, you copy the current list (list) into the result:

result.add(ArrayList(list)) → takes O(n) time.

So total work =
Number of permutations × cost of copying one permutation
→ n! × n → O(n × n!)

*/
    fun permute(nums: IntArray): List<List<Int>> {
        val result = mutableListOf<List<Int>>()
        val list = mutableListOf<Int>()
        val used = BooleanArray(nums.size)
        permuteHelper(nums, used, list, result)
        return result
    }

    fun permuteHelper(
        nums: IntArray,
        used: BooleanArray,
        list: MutableList<Int>,
        result: MutableList<List<Int>>
    ) {
        if (list.size == nums.size) {
            result.add(ArrayList(list))
            return
        }

        for (i in nums.indices) {
            if (used[i]) continue
            used[i] = true
            list.add(nums[i])
            permuteHelper(nums, used, list, result)
            used[i] = false
            list.removeAt(list.size - 1)
        }
    }
}

fun main() {
    val obj = Permutations()
    println(obj.permute(intArrayOf(1, 2, 3)))
}